From eac9b00d9f985ccdc30a8a510119cb76e40d9ae6 Mon Sep 17 00:00:00 2001
From: Brian Behlendorf <behlendorf1@llnl.gov>
Date: Fri, 6 Nov 2015 10:10:27 -0800
Subject: [PATCH] Retry fetching instances after creation

There appears to exist a window where an instance has been successfully
created but it does not yet appear in the instance list.  To avoid
spurious failures retry in the case of this exception.
---
 master/buildbot/buildslave/ec2.py | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/master/buildbot/buildslave/ec2.py b/master/buildbot/buildslave/ec2.py
index 84de44d..4a59bce 100644
--- a/master/buildbot/buildslave/ec2.py
+++ b/master/buildbot/buildslave/ec2.py
@@ -384,7 +384,21 @@ class EC2LatentBuildSlave(AbstractLatentBuildSlave):
             instance_id = request.instance_id
             reservations = self.conn.get_all_instances(instance_ids=[instance_id])
             self.instance = reservations[0].instances[0]
-            instance_id, image_id, start_time = self._wait_for_instance(self.get_image())
+            attempts = 0
+            while attempts < 10:
+                try:
+                    instance_id, image_id, start_time = self._wait_for_instance(self.get_image())
+                    break
+                except boto.exception.BotoSeverError, e:
+                    attempts += 1
+                    log.msg('%s %s failed to find instance %s (%s)retry %d' %
+                            (self.__class__.__name__, self.slavename,
+                             self.instance.id, self.instance.state, retry))
+                    if e.error_code == 'InvalidInstanceID.NotFound':
+                         time.sleep(1)
+                         continue
+                    else:
+                         raise
             return instance_id, image_id, start_time, True
 
     def _request_spot_instance(self):
-- 
1.9.1

